cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemoteApplicationDetailFetcher.java

```diff
@@ -0,0 +1,99 @@
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+import io.cdap.common.http.ObjectResponse;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.List;
+
+/**
+ * Fetch application detail via REST API calls
+ */
+public class RemoteApplicationDetailFetcher implements AbstractApplicationDetailFetcher {
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder()).create();
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemoteApplicationDetailFetcher(final DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false), Constants.Gateway.API_VERSION_3);
+  }
+
+  /**
+   * Get the application detail for the given application id
+   */
+  public ApplicationDetail get(ApplicationId appId) throws IOException, NotFoundException {
+    String url = String.format("namespaces/%s/apps/%s/versions/%s",
+                               appId.getNamespace(), appId.getApplication(), appId.getVersion());
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse httpResponse;
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(appId);
+    }
+    return GSON.fromJson(httpResponse.getResponseBodyAsString(), ApplicationDetail.class);
+  }
+
+  /**
+   * Get details of all applications in the given namespace
+   */
+  public List<ApplicationDetail> list(String namespace) throws IOException, NotFoundException {
+    String url = String.format("namespaces/%s/apps", namespace);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    HttpResponse httpResponse;
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(new NamespaceId(namespace));
+    }
+    ObjectResponse<List<ApplicationDetail>> objectResponse = ObjectResponse.fromJsonBody(
+      httpResponse, new TypeToken<List<ApplicationDetail>>() {
+      }.getType(), GSON);
+    return objectResponse.getResponseObject();
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException, NotFoundException {
+    HttpResponse httpResponse = remoteClient.execute(request);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException("Not found");
+    }
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      throw new IOException(String.format("Request failed %s", httpResponse.getResponseBodyAsString()));
```

Comment:
just use the response body by itself without the 'Request failed', otherwise the message can end up looking weird.

Reply:
Fixed

Reply:
This isn't fixed either, is there a missing commit?

Reply:
I think I might have missed committing that change. 

Fixed. Please double check.